### PR TITLE
For segrep enabled indices, use NRTReplicationEngine for replica shards

### PR DESCRIPTION
**Pull request by dreamer-89**
Signed-off-by: Suraj Singh <surajrider@gmail.com>

### Description
Update EngineFactory to return NRTReplicationEngine for SegRep enabled replica shards. Today both primary & replica uses [ReplicationEngine](https://github.com/opensearch-project/cross-cluster-replication/blob/28d24507a6c641507ac44e32a353ad33eef166aa/src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt#L326) (extends InternalEngine from OpenSearch) though [NRTReplicationEngine](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java#L49) is needed on replica shards for SegmentReplication to work. 

### Resolves
https://github.com/opensearch-project/OpenSearch/issues/3823

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt [hunk 1]
```diff
@@ -359,7 +363,13 @@ internal class ReplicationPlugin : Plugin(), ActionPlugin, PersistentTaskPlugin,
 
     override fun getEngineFactory(indexSettings: IndexSettings): Optional<EngineFactory> {
         return if (indexSettings.settings.get(REPLICATED_INDEX_SETTING.key) != null) {
-            Optional.of(EngineFactory { config -> ReplicationEngine(config) })
+            Optional.of(EngineFactory { config ->
+                if (indexSettings.settings.get(INDEX_REPLICATION_TYPE_SETTING.key).equals("SEGMENT")) {
+                    if (config.isReadOnlyReplica) NRTReplicationEngine(config) else ReplicationEngine(config)
```
Comment (ankitkala) [inline src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt:368]: It should be `indexSettings.settings.get(INDEX_REPLICATION_TYPE_SETTING.key).equals("SEGMENT") && config.isReadOnlyReplica`

Comment (ankitkala) [reply src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt:368]: Actually, we shouldn't be setting the Engine here. Instead, we can override the ReplicationEngine only if the shard is `primary`. This way the we don't replicate the logic and  existing `EngineFactory` would kick in add the `NRTReplciationEngine` for replicas.

Comment (ankitkala) [reply src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt:368]: Just realised that primary shard check might not work as we load the `EngineFactory` as index creation level and not shard. 
We'll need to keep the logic to load the `NRTReplicationEngine` here.

Comment (dreamer-89) [reply src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt]: Updated the logic here.

**Review by ankitkala — APPROVED** [review]

File: src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt [hunk 2]
```diff
@@ -359,7 +363,14 @@ internal class ReplicationPlugin : Plugin(), ActionPlugin, PersistentTaskPlugin,
 
     override fun getEngineFactory(indexSettings: IndexSettings): Optional<EngineFactory> {
         return if (indexSettings.settings.get(REPLICATED_INDEX_SETTING.key) != null) {
-            Optional.of(EngineFactory { config -> ReplicationEngine(config) })
+            Optional.of(EngineFactory { config ->
+                // Use NRTSegmentReplicationEngine for SEGMENT replication type indices replica shards
+                if (config.isReadOnlyReplica && indexSettings.settings.get(INDEX_REPLICATION_TYPE_SETTING.key) != null && indexSettings.settings.get(INDEX_REPLICATION_TYPE_SETTING.key).equals("SEGMENT")) {
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt:368]: - we can drop explicit `null` check and use "?" instead for obtaining replication type setting. 
- Can we use constant for "SEGMENT" (from the allowed values)?

Comment (dreamer-89) [reply src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt]: Looks like `config.isReadOnlyReplica` already have segrep settings check inside EngineConfig. Removing segrep setting explicit check. 

```
    public boolean isReadOnlyReplica() {
        return indexSettings.isSegRepEnabled() && isReadOnlyReplica;
    }
```

**Review by ankitkala — APPROVED** [review]

**Review by saikaranam-amazon — APPROVED** [review]

**Pull request merged**